pkg/interface/src/logic/lib/useStorage.ts

```diff
@@ -1,41 +1,57 @@
-import { useCallback, useMemo, useEffect, useRef, useState } from "react";
-import { S3State } from "../../types/s3-update";
+import {useCallback, useMemo, useEffect, useRef, useState} from 'react';
+import {S3State} from '../../types/s3-update';
+import {GcpState} from '../../types/gcp-state';
 import S3 from "aws-sdk/clients/s3";
-import { dateToDa, deSig } from "./util";
+import GcpClient from './GcpClient';
+import {StorageClient, StorageAcl} from './StorageClient';
+import {dateToDa, deSig} from "./util";
 
-export interface IuseS3 {
+
+export interface IuseStorage {
   canUpload: boolean;
   upload: (file: File, bucket: string) => Promise<string>;
   uploadDefault: (file: File) => Promise<string>;
   uploading: boolean;
   promptUpload: () => Promise<string | undefined>;
 }
 
-const useS3 = (s3: S3State, { accept = '*' } = { accept: '*' }): IuseS3 => {
+const useStorage = (s3: S3State, gcp: GcpState,
+                    { accept = '*' } = { accept: '*' }): IuseStorage => {
   const [uploading, setUploading] = useState(false);
 
-  const client = useRef<S3 | null>(null);
+  const client = useRef<StorageClient | null>(null);
 
   useEffect(() => {
-    if (!s3.credentials) {
-      return;
+    // prefer GCP if available, else use S3.
+    if (gcp.token !== undefined) {
+      client.current = new GcpClient(gcp.token.accessKey);
+    } else {
+      // XX ships currently always have S3 credentials, but the fields are all
```

Comment:
I see. Well, empty strings do evaluate to false!

Reply:
JavaScript!